### ISSUE TITLE
Full specified names UUIDs to be seeded from their concept's UUID.

### ISSUE THREAD

**Issue by mks-d** [mekomsolutions/openmrs-module-initializer#141]
#### Background
See
* OpenMRS Talk: '[How concept names are breaking dbsync](https://talk.openmrs.org/t/how-concept-names-are-breaking-dbsync/34776?u=mksd)'
* This discussion on Slack: https://openmrs.slack.com/archives/C0178G3A1T7/p1632834262050900

#### Proposed Solution
By default fully specified names should be set a UUID that can be seeded and inferred from their concept's UUID. This would happen around [here](https://github.com/mekomsolutions/openmrs-module-initializer/blob/f15f65ece98374487c2596cd8ad8f3e479b7eb0c/api/src/main/java/org/openmrs/module/initializer/api/c/ConceptLineProcessor.java#L61-L62) and go along the lines of:
```java
import java.util.UUID;
...
String seed = concept.getUuid() + "_" + locale.toString();
conceptName.setUuid(UUID.nameUUIDFromBytes(seed.getBytes()).toString());
```
This would ensure that, in each locale, each FSN is always set a UUID in the exact same and predictable way.

Credits to @rbuisson for sharing his examples [here](https://github.com/rbuisson/c2c-migration/blob/b186aa0466463f3ecbebefaee0c9754d704c5d1d/src/main/java/net/mekomsolutions/c2c/migration/converter/PatientConverter.java#L57) and [here](https://github.com/rbuisson/c2c-migration/blob/b186aa0466463f3ecbebefaee0c9754d704c5d1d/src/main/java/net/mekomsolutions/c2c/migration/converter/PatientConverter.java#L77-L78).

#### Acceptance Criteria
This enhancement should come with a Liquibase changeset that applies the above seeding logic to existing FSNs. It is possible to apply Java-based logic through a changeset, see [`<customChange/>`](https://docs.liquibase.com/change-types/community/custom-change.html).

**Comment by mks-d** [comment]
>Like @mseaton I'd be a bit hesitant to apply something like that to an existing database and probably not as a default.

I'm curious to know of practical cases that would forbid this, @mseaton @ibacher could you share some?

**Comment by mseaton** [comment]
> I'm curious to know of practical cases that would forbid this, @mseaton @ibacher could you share some?

The sync module is the place where I'd most be concerned about this, but that same issue may present elsewhere - namely that concept names are identified by uuids in some other serialized data representation or integration, and changing them under the hood may impact these tools.  That said, anyone already using Iniz for managing concepts is already in a situation where concept names uuid may change under the hood on them at any time, as they are recreated with random uuids anytime the concept is edited, so for those implementations this likely is fine.

**Comment by mseaton** [comment]
> If we limit this to FSNs (and that's the intent), the pair name as string + locale should be a unique seed no?

Per the above comments, I just think if we're going to go to the trouble to implement this, we should not assume it will just be used for FSNs, but we should make it robust enough to use for any concept names that exist in the dictionary.

Unfortunately the problem with that is that it makes it impossible to change any details about the concept name after it is created, and retain the same generated uuid.  So it means that changing a concept from a FSN to a Synonym, or changing the locale (to, say "en_US" from "en") would result in a different uuid, rather than an edit to an existing uuid.  I assume this is why you have the suggestion that you have here to use just the concept uuid in the seed.  Hmm..

**Comment by mseaton** [comment]
No, because not everyone uses Iniz to manage their concepts, and we don't want those who upgrade Iniz to have their concept names changed under them.

What about an approach that:

1. Updating the code so that if a FSN is not specified with a UUID, generate the uuid using the algorithm proposed, and then
2. Create a custom liquibase changeset that simply deletes the existing checksums under the Concepts domain.

This will force all Concepts to be reloaded that are managed via CSVs, and will ensure that their names are recreated using the new algorithm.

@mks-d ?

**Comment by mks-d** [comment]
@mseaton brilliant idea, thanks.

**Comment by mseaton** [comment]
@Ruhanga - this issue should be closed, correct?  This has been completed and merged?  @mks-d ?
